File: src/python/xerxes/audit/otel_exporter.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Translate :class:`AuditEvent` objects into OpenTelemetry spans.

Each turn becomes a parent span (``xerxes.turn``); tool attempts,
completions, failures, skill uses, and errors are attached to that
span as OpenTelemetry events when the turn is in flight, otherwise
emitted as short standalone spans. When ``opentelemetry`` is not
installed the collector falls back to an in-memory dict log accessible
via :attr:`OTelCollector.fallback_log` for tests.
"""

from __future__ import annotations

import logging
import typing as tp
from contextlib import contextmanager
from datetime import datetime

from .events import (
    AuditEvent,
    ErrorEvent,
    SkillUsedEvent,
    ToolCallAttemptEvent,
    ToolCallCompleteEvent,
    ToolCallFailureEvent,
    TurnEndEvent,
    TurnStartEvent,
)

logger = logging.getLogger(__name__)


def _try_import_otel() -> tuple[tp.Any, tp.Any] | None:
    """Return ``(trace, tracer)`` if OpenTelemetry imports cleanly, else ``None``."""
    try:
        from opentelemetry import trace as _trace
    except ImportError:
        return None
    try:
        return _trace, _trace.get_tracer("xerxes")
    except Exception:
        return None


class OTelCollector:
    """Audit collector backed by OpenTelemetry tracing.

    Each ``TurnStartEvent`` opens a span; intermediate events attach as
    span events; ``TurnEndEvent`` closes the span. Without OTel the
    collector buffers entries in :attr:`fallback_log` instead.
    """

    def __init__(self, service_name: str = "xerxes") -> None:
        """Construct the collector and try to acquire a tracer.

        ``service_name`` is attached to every span as the ``service.name``
        attribute and used by downstream OTel collectors for grouping.
        """
        self.service_name = service_name
        otel = _try_import_otel()
        if otel is None:
            self._trace = None
            self._tracer = None
            logger.info("OpenTelemetry not installed; OTelCollector will log instead")
        else:
            self._trace, self._tracer = otel
        self._open_turn_spans: dict[str, tp.Any] = {}
        self._noop_log: list[dict[str, tp.Any]] = []

    def emit(self, event: AuditEvent) -> None:
        """Dispatch ``event`` to its type-specific handler.

        Unknown event types are recorded with a generic name. Exceptions
        from the OTel SDK are caught and logged so audit failures never
        propagate to the caller.
        """
        try:
            if isinstance(event, TurnStartEvent):
                self._on_turn_start(event)
            elif isinstance(event, TurnEndEvent):
                self._on_turn_end(event)
            elif isinstance(event, ToolCallAttemptEvent):
                self._on_tool_attempt(event)
            elif isinstance(event, ToolCallCompleteEvent):
                self._on_tool_complete(event)
            elif isinstance(event, ToolCallFailureEvent):
                self._on_tool_failure(event)
            elif isinstance(event, SkillUsedEvent):
                self._on_skill_used(event)
            elif isinstance(event, ErrorEvent):
                self._on_error(event)
            else:
                self._record_event(event.event_type, event)
        except Exception:
            logger.warning("OTelCollector failed to handle %s", event.event_type, exc_info=True)

    def flush(self) -> None:
        """End every still-open turn span (used at shutdown)."""
        for span in list(self._open_turn_spans.values()):
            try:
                span.end()
            except Exception:
                pass
        self._open_turn_spans.clear()

    @property
    def has_otel(self) -> bool:
        """Return ``True`` when an OpenTelemetry tracer is available."""
        return self._tracer is not None

    @property
    def fallback_log(self) -> list[dict[str, tp.Any]]:
        """Return a copy of the in-memory log used when OTel is missing."""
        return list(self._noop_log)

    @contextmanager
    def _span(self, name: str, attributes: dict[str, tp.Any]):
        """Yield a short-lived span or append a fallback log entry."""
        if self._tracer is None:
            self._noop_log.append({"name": name, "attributes": dict(attributes)})
            yield None
            return
        span = self._tracer.start_span(name=name, attributes=_clean_attrs(attributes))
        try:
            yield span
        finally:
            try:
                span.end()
            except Exception:
                pass

    def _record_event(self, name: str, event: AuditEvent) -> None:
        """Attach ``event`` to its turn span, or emit a standalone span."""
        attrs = _clean_attrs(event.to_dict())
        turn_id = getattr(event, "turn_id", None)
        if turn_id and turn_id in self._open_turn_spans and self._tracer is not None:
            try:
                self._open_turn_spans[turn_id].add_event(name, attributes=attrs)
                return
            except Exception:
                pass
        if self._tracer is None:
            self._noop_log.append({"name": name, "attributes": attrs})
        else:
            with self._span(name, attrs):
                pass

    def _on_turn_start(self, event: TurnStartEvent) -> None:
        """Open and remember a ``xerxes.turn`` span keyed by ``turn_id``."""
        if self._tracer is None or not event.turn_id:
            self._noop_log.append({"name": "turn", "attributes": _clean_attrs(event.to_dict())})
            return
        span = self._tracer.start_span(
            name="xerxes.turn",
            attributes=_clean_attrs(
                {
                    "xerxes.turn_id": event.turn_id,
                    "xerxes.agent_id": event.agent_id,
                    "xerxes.session_id": event.session_id,
                    "xerxes.prompt_preview": event.prompt_preview,
                    "service.name": self.service_name,
                }
            ),
        )
        self._open_turn_spans[event.turn_id] = span

    def _on_turn_end(self, event: TurnEndEvent) -> None:
        """Close the turn span and record the final function-call count."""
        if not event.turn_id:
            return
        span = self._open_turn_spans.pop(event.turn_id, None)
        if span is None:
            return
        try:
            span.set_attribute("xerxes.function_calls_count", event.function_calls_count)
            span.end()
        except Exception:
            pass

    def _on_tool_attempt(self, event: ToolCallAttemptEvent) -> None:
        """Record a ``tool.attempt:<name>`` event."""
        self._record_event(f"tool.attempt:{event.tool_name}", event)

    def _on_tool_complete(self, event: ToolCallCompleteEvent) -> None:
        """Record a ``tool.complete:<name>`` event."""
        self._record_event(f"tool.complete:{event.tool_name}", event)

    def _on_tool_failure(self, event: ToolCallFailureEvent) -> None:
        """Record a ``tool.failure:<name>`` event."""
        self._record_event(f"tool.failure:{event.tool_name}", event)

    def _on_skill_used(self, event: SkillUsedEvent) -> None:
        """Record a ``skill.used:<name>`` event."""
        self._record_event(f"skill.used:{event.skill_name}", event)

    def _on_error(self, event: ErrorEvent) -> None:
        """Record an ``error:<type>`` event."""
        self._record_event(f"error:{event.error_type}", event)


def _clean_attrs(d: dict[str, tp.Any]) -> dict[str, tp.Any]:
    """Coerce ``d`` to OpenTelemetry-safe attribute values.

    Drops ``None``, stringifies non-primitives (truncated to 200 chars),
    and renders datetimes as ISO strings. The OTel SDK only accepts
    primitive scalar values; objects survive only as their ``str`` form.
    """
    out: dict[str, tp.Any] = {}
    for k, v in d.items():
        if v is None:
            continue
        if isinstance(v, str | int | float | bool):
            out[k] = v
        elif isinstance(v, datetime):
            out[k] = v.isoformat()
        else:
            out[k] = str(v)[:200]
    return out


__all__ = ["OTelCollector"]

File: xerxes/src/audit/otelExporter.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import {
  type AuditEvent,
  ErrorEvent,
  SkillUsedEvent,
  ToolCallAttemptEvent,
  ToolCallCompleteEvent,
  ToolCallFailureEvent,
  TurnEndEvent,
  TurnStartEvent,
} from './events.js'
import type { AuditCollector } from './collector.js'

/** Scalar values accepted by the portion of the OpenTelemetry API used by this adapter. */
export type OTelAttributeValue = boolean | number | string
export type OTelAttributes = Readonly<Record<string, OTelAttributeValue>>

/** Structural subset of an OpenTelemetry span, keeping the runtime free of a hard package dependency. */
export interface OTelSpan {
  addEvent(name: string, attributes?: OTelAttributes): void
  end(): void
  setAttribute(key: string, value: OTelAttributeValue): void
}

/** Structural subset of an OpenTelemetry tracer accepted through dependency injection. */
export interface OTelTracer {
  startSpan(name: string, options?: Readonly<{ attributes?: OTelAttributes }>): OTelSpan
}

export interface OTelCollectorOptions {
  readonly serviceName?: string
  /** An OpenTelemetry tracer supplied by the application, if OpenTelemetry is installed. */
  readonly tracer?: OTelTracer
  /** Maximum turn spans held open at once; the oldest is ended and evicted beyond this. */
  readonly maxOpenTurns?: number
  /** Maximum retained no-op fallback entries; the oldest are dropped beyond this. */
  readonly maxFallbackEntries?: number
}

/** A recorded no-op span/event used when no tracer has been injected. */
export interface OTelFallbackEntry {
  readonly attributes: OTelAttributes
  readonly name: string
}

/**
 * Turns audit records into OpenTelemetry spans/events when supplied a tracer.
 *
 * The collector deliberately does not import OpenTelemetry itself. Applications can inject
 * their configured tracer, while Bun-only deployments retain an inspectable no-op log.
 */
export class OTelCollector implements AuditCollector {
  static readonly DEFAULT_MAX_OPEN_TURNS = 1_000
  static readonly DEFAULT_MAX_FALLBACK_ENTRIES = 10_000

  private readonly noopLog: OTelFallbackEntry[] = []
  private readonly openTurnSpans = new Map<string, OTelSpan>()
  private readonly maxOpenTurns: number
  private readonly maxFallbackEntries: number
  private evictedTurns = 0
  private evictedFallbacks = 0
  readonly serviceName: string
  readonly tracer: OTelTracer | undefined

  constructor(options?: OTelCollectorOptions)
  constructor(serviceName?: string, tracer?: OTelTracer)
  constructor(optionsOrServiceName: OTelCollectorOptions | string = {}, tracer?: OTelTracer) {
    if (typeof optionsOrServiceName === 'string') {
      this.serviceName = optionsOrServiceName
      this.tracer = tracer
      this.maxOpenTurns = OTelCollector.DEFAULT_MAX_OPEN_TURNS
      this.maxFallbackEntries = OTelCollector.DEFAULT_MAX_FALLBACK_ENTRIES
      return
    }
    this.serviceName = optionsOrServiceName.serviceName ?? 'xerxes'
    this.tracer = optionsOrServiceName.tracer
    this.maxOpenTurns = boundedSize(optionsOrServiceName.maxOpenTurns, OTelCollector.DEFAULT_MAX_OPEN_TURNS, 'maxOpenTurns')
    this.maxFallbackEntries = boundedSize(optionsOrServiceName.maxFallbackEntries, OTelCollector.DEFAULT_MAX_FALLBACK_ENTRIES, 'maxFallbackEntries')
  }

  emit(event: AuditEvent): void {
    try {
      if (event instanceof TurnStartEvent) {
        this.onTurnStart(event)
        return
      }
      if (event instanceof TurnEndEvent) {
        this.onTurnEnd(event)
        return
      }
      if (event instanceof ToolCallAttemptEvent) {
        this.recordEvent(`tool.attempt:${event.toolName}`, event)
        return
      }
      if (event instanceof ToolCallCompleteEvent) {
        this.recordEvent(`tool.complete:${event.toolName}`, event)
        return
      }
      if (event instanceof ToolCallFailureEvent) {
        this.recordEvent(`tool.failure:${event.toolName}`, event)
        return
      }
      if (event instanceof SkillUsedEvent) {
        this.recordEvent(`skill.used:${event.skillName}`, event)
        return
      }
      if (event instanceof ErrorEvent) {
        this.recordEvent(`error:${event.errorType}`, event)
        return
      }
      this.recordEvent(event.eventType, event)
    } catch {
      // Audit exports must never prevent agent work. Keep enough context for local diagnostics.
      this.recordFallback('audit.export_error', cleanOtelAttributes(event.toRecord()))
    }
  }

  flush(): void {
    for (const span of this.openTurnSpans.values()) {
      try {
        span.end()
      } catch {
        // Best effort shutdown mirrors the behavior of external telemetry SDKs.
      }
    }
    this.openTurnSpans.clear()
  }

  get hasOtel(): boolean {
    return this.tracer !== undefined
  }

  /** A copy of no-op entries, so callers cannot mutate the collector's diagnostics. */
  get fallbackLog(): OTelFallbackEntry[] {
    return this.noopLog.map(entry => ({ name: entry.name, attributes: { ...entry.attributes } }))
  }

  get openTurnCount(): number {
    return this.openTurnSpans.size
  }

  /** Turn spans evicted because cancelled turns never produced a TurnEnd event. */
  get evictedTurnCount(): number {
    return this.evictedTurns
  }

  /** No-op fallback entries dropped to keep the diagnostic log bounded. */
  get evictedFallbackCount(): number {
    return this.evictedFallbacks
  }

  private onTurnStart(event: TurnStartEvent): void {
    const attributes = cleanOtelAttributes({
      'xerxes.turn_id': event.turnId,
      'xerxes.agent_id': event.agentId,
      'xerxes.session_id': event.sessionId,
      'xerxes.prompt_preview': event.promptPreview,
      'service.name': this.serviceName,
    })
    if (this.tracer === undefined || event.turnId === undefined) {
      this.recordFallback('turn', attributes)
      return
    }
    // Refresh recency so long-lived turns are not evicted ahead of stale ones.
    this.openTurnSpans.delete(event.turnId)
    while (this.openTurnSpans.size >= this.maxOpenTurns) {
      const oldest = this.openTurnSpans.keys().next().value
      if (oldest === undefined) break
      const stale = this.openTurnSpans.get(oldest)
      this.openTurnSpans.delete(oldest)
      this.evictedTurns += 1
      try {
        stale?.end()
      } catch {
        // Ending an evicted span is best effort.
      }
    }
    this.openTurnSpans.set(event.turnId, this.tracer.startSpan('xerxes.turn', { attributes }))
  }

  private onTurnEnd(event: TurnEndEvent): void {
    if (event.turnId === undefined) return
    const span = this.openTurnSpans.get(event.turnId)
    if (span === undefined) return
    this.openTurnSpans.delete(event.turnId)
    try {
      span.setAttribute('xerxes.function_calls_count', event.functionCallsCount)
      span.end()
    } catch {
      // Do not leak a telemetry failure into the agent loop.
    }
  }

  private recordEvent(name: string, event: AuditEvent): void {
    const attributes = cleanOtelAttributes(event.toRecord())
    const turnId = event.turnId
    const turnSpan = turnId === undefined ? undefined : this.openTurnSpans.get(turnId)
    if (this.tracer !== undefined && turnSpan !== undefined) {
      turnSpan.addEvent(name, attributes)
      return
    }
    if (this.tracer === undefined) {
      this.recordFallback(name, attributes)
      return
    }
    const span = this.tracer.startSpan(name, { attributes })
    try {
      // The short span itself carries all attributes.
    } finally {
      span.end()
    }
  }

  private recordFallback(name: string, attributes: OTelAttributes): void {
    this.noopLog.push({ name, attributes: { ...attributes } })
    const overflow = this.noopLog.length - this.maxFallbackEntries
    if (overflow > 0) {
      this.noopLog.splice(0, overflow)
      this.evictedFallbacks += overflow
    }
  }
}

function boundedSize(value: number | undefined, fallback: number, name: string): number {
  if (value === undefined) return fallback
  if (!Number.isInteger(value) || value <= 0) {
    throw new Error(`OTelCollector ${name} must be a positive integer`)
  }
  return value
}

/** Coerce a JSON record into OTel-safe scalar attributes and remove absent values. */
export function cleanOtelAttributes(values: Readonly<Record<string, unknown>>): OTelAttributes {
  const attributes: Record<string, OTelAttributeValue> = {}
  for (const [key, value] of Object.entries(values)) {
    if (value === undefined || value === null) continue
    if (typeof value === 'string' || typeof value === 'boolean') {
      attributes[key] = value
      continue
    }
    if (typeof value === 'number') {
      attributes[key] = Number.isFinite(value) ? value : String(value)
      continue
    }
    if (value instanceof Date) {
      attributes[key] = value.toISOString()
      continue
    }
    attributes[key] = stringifyAttribute(value)
  }
  return attributes
}

function stringifyAttribute(value: unknown): string {
  try {
    const rendered = JSON.stringify(value)
    if (rendered !== undefined) return rendered.slice(0, 200)
  } catch {
    // Fall through to String for cyclic or host objects.
  }
  return String(value).slice(0, 200)
}
